### ISSUE TITLE
Cannot Import __main__

### ISSUE THREAD

**Issue by crantila** [pypyjs/pypyjs#161]
My quest is to make the ``unittest`` module work. It was pretty easy to remove the dependency on the ``signals`` module, but before I submit a pull request I want to run the ``unittest`` module's unit tests. And that's how I happened upon that module's use of ``import __main__``.

I considered rewriting to avoid importing ``__main__``, but that's actually a very clever way to do something otherwise very difficult, and since ``__main__`` will always already be available to PyPy.js, it shouldn't be too hard to figure out how to get the ``import __main__`` call to work properly, right? Probably a good introduction to how the import mechanism works, at least.

So if you have any ideas how to approach this, I'd be happy to hear them. Otherwise... expect questions or a resolution in the near future!

**Comment by rfk** [comment]
Interesting.  We sort-of already create a stub `__main__` module here at [1], in which to execute top-level statements.  It may be as simple as just inserting that object into `sys.modules["__main__"]` as part of the startup procedure.

[1] https://github.com/pypyjs/pypyjs/blob/master/lib/pypyjs.js#L379

**Comment by crantila** [comment]
As it turns out, that's what to do. Near as I can tell, it's also what CPython and PyPy do. I even tried a bunch of other strategies, to see if I could find something that would be preferrable for some reason. Nope.

So line 379ff becomes:

    "top_level_scope = {'__name__': '__main__'}",
    "sys.modules['__main__'] = top_level_scope"

Shall I prepare a pull request, or would you like to incorporate this ridiculously simple solution? I don't mind either way.

**Comment by crantila** [comment]
No, I lied, that doesn't work either. It has the same problem as everything else: easy enough to have a ``__main__`` that's a dict, but how to make it a module? Back to work...

**Comment by perkinslr** [comment]
Not sure exactly, but this may get you started.  At least it works in CPython...

    from types import ModuleType
    main = ModuleType('__main__')
    main.__dict__.update(top_level_scope)
    sys.modules['__main__']=main

**Comment by crantila** [comment]
I already discovered a similar hack, setting ``__dict__`` on a purpose-built class. I don't think a dict's update() function will work in this case, because the "update()d" dict wouldn't receive future updates without some convoluted system.

But ``main.__dict__ = top_level_scope``? Still investigating; I'll have a unit test too!

**Comment by perkinslr** [comment]
The `types` module avoids you having to do the `type(lambda: None)` hackery yourself, but that's basically how it gets most of the type references.  (`ModuleType=type(sys)` for example).  `main.__dict__ =` won't work because `__dict__` is a read-only property on modules.  After calling `main.__dict__.update`, then set `top_level_scope=main.__dict__`, that way all the references should point to the same object.

**Comment by crantila** [comment]
Doesn't work. I'll come back to this tomorrow. Thanks for your help so far!

**Comment by perkinslr** [comment]
Drat, might be able to do something by subclassing `ModuleType`, the `__dict__` property will still be read-only, but it would let you hijack `__getattribute__` and `__setattribute__` and redirect them to `top_level_scope.__[get/set]__item`

**Comment by pyhedgehog** [comment]
Maybe better to start with creating `module`?
```Python
main = type(sys)('__main__')
sys.modules['__main__'] = main
top_level_scope = main.__dict__
```

**Comment by perkinslr** [comment]
I suspect that will have the same issue that his last attempt had (but I'm not certain, since he didn't say what that was).  It also may be due to the missing attributes when you dynamically create a module.

    >>> import __main__
    >>> vars(__main__)
    {'__builtins__': <module '__builtin__' (built-in)>, '__name__': '__main__', '__main__': <module '__main__' (built-in)>, '__doc__': None, '__package__': None}

    >>> vars(ModuleType('__main__'))
    {'__name__': None, '__doc__': None}

Probably need to provide the missing attributes.

**Comment by rfk** [comment]
I think @pyhedgehog is on the right track here, we should start with creating a proper `__main__` module and then have the `pypyjs.js` shell code execute stuff in its scope.  There's nothing special about `top_level_scope`, it's just a place for the `pypyjs.js` shell code to store its global state, so we should be able to replace all uses of it with `__main__.__dict__`.

**Comment by crantila** [comment]
Oh... handy... I was just coming here to tell you that we'll have to change things significantly in order to get this to work. So you're okay with me making ``__main__`` into a module, and changing all the relevant functions so they work with a module instead of a dict?

**Comment by crantila** [comment]
I've tentatively got it working. You can see the commit just above this message.

I'd like to do some further testing, to make sure it works for what I actually wanted to use it for, but I also thought I would let you see the changes as soon as possible. Please feel free to comment and criticize&mdash;I'm still quite new to JavaScript, and this is also an unusual area of Python for me!

**Comment by rfk** [comment]
Fixed by #165

**Comment by crantila** [comment]
Could we re-open this? I just discovered that my PR doesn't work with ``repl()``. The fix is easy, but I'd like to add a test for it, which might take a day or two, especially if @albertjan's move to ES6 happens first. Thanks!

**Comment by rfk** [comment]
re-opened

**Comment by albertjan** [comment]
@crantila that bug is fixed in my es6 branch

**Comment by crantila** [comment]
@rfk is it worth fixing this properly, since @albertjan already fixed it on his ES6 branch? Will his PR be merged before the next release, making my own fix redundant?
